04-onion-routing.md

```diff
@@ -367,6 +368,149 @@ otherwise meets the amount criterion (eg. some other failure, or
 invoice timeout), however if it were to fulfill only some of them,
 intermediary nodes could simply claim the remaining ones.
 
+### Onion Messages
+
+Onion messages have an onion with an alternate `hop_payload`
+format: a `bigsize` followed by a `onionmsg_payload`.  Note that there
+is no legacy format, thus a `bigsize` of 0 means no payload.
+
+An intermediate node expects an `enctlv` which it can decrypt using
+the `blinding` which it is handed along with the onion message.
+
+1. `tlv_stream`: `onionmsg_payload`
```

Comment:
I think I mentioned this off GitHub, but I really think there needs to be a separate TLV stream for the actual message content. Shoving it in the same one is cool, but in practice if we have multiple message types we're gonna want to pass a whole TLV stream to different parsers, not have each message type hook the same TLV stream parsing logic.

Reply:
You mean we have a `payload` blob of bytes at this level, and YA tlv inside it?  Possible, but I don't think demux is going to get any prettier?

Reply:
That's one way to accomplish it, yes. I don't really understand how to demux without some second-level thing here - we have a tlv stream that contains fields that the onion message decoder/processor needs to handle, and fields that the onion-message-type handler needs to process. Presumably those need to be two separate things because there could be different types of onion messages. Arguably there needs to be, per type of onion message, a different, single TLV record in the payload here that declares the type and has all the type-specific data (eg type 1 would be "BOLT 12 offer request" and type 3 would be "BOLT 12 offer response", and....etc).

Reply:
Perhaps we just say "64 and above"?  It is much nice to use one number per message type, IMHO, rather than one per feature.

Reply:
SGTM, let's keep it like this for now and we can have this discussion when we review features that actually add tlvs for the final node.